public/app/features/datasources/components/EditDataSource.tsx

```diff
@@ -162,6 +163,12 @@ export function EditDataSourceView({
         disabled={readOnly || !hasWriteRights}
       />
 
+      <DataSourceDescription
+        dataSourceName={dataSourceMeta.name}
+        docsLink={dataSourceMeta.info?.links?.find((l) => l.name === 'Docs' || l.name === 'Documentation')?.url}
+        hasRequiredFields={dataSourceMeta.hasRequiredConfigFields}
+      />
```

Comment:
This to me feels strangely placed / wrong way around. Since each data source is responsible for config via the ConfigEditor, it feels like this is better done via a react component that data sources can opt to use in ConfigEditor (instead of property in plugin.json)

Reply:
thoughts @mckn @leventebalogh ? 

Best thing would be a kind of declarative option / data source settings framework that data sources could use (similar to panel settings)  :)

Reply:
Initially we wanted every datasource to import `DataSourceDescription` component and add it to its config editor. But then seems like it will be duplicated in every data source, so we decided to add it to core to avoid duplication and inconsistencies. If you think there is a better approach, please advise.

cc @catherineymgui @yesoreyeram

Reply:
+1 with @torkelo . This looks like something each datasource should render on its own config editor. 

@cletter7 you might be thinking only on datasources made by us where we might reuse this code? Keep in mind plugins are developed also by the community and the code is different on each plugin. What if you create a grafana/ui component that has the style and the format you want and accept the text? this way you avoid code duplication and keep the consistency you want?

Reply:
Yeah, I would vote for using React to render these from the plugin directly, too. 

With the current solution we would have two different ways to configure what is displayed on the custom config page for plugins: one using config editor, and one by using certain fields of the plugin.json, which is not ideal. Exposing this component via `@grafana/ui` could be an option, however as it currently doesn't hold too much logic, I am not sure it makes sense. 

We could probably make marking / displaying required fields easier and the error handling better? Also, as for linking to documentation, could we just link to the GCOM page of the plugins, which shows the README? I think in most cases it is the READMEs that hold information about how to configure the plugins.

Reply:
I understand there are other plugins. By default if they have no docs link only this will be rendered on top:
```
Before you can use the {dataSourceName} data source, you must configure it below or in the config file.
```
Is there a way to target internal plugins only?

> This looks like something each datasource should render on its own config editor.

If each plugin does this, why to not to render it in one place? Maybe there is a reason, just want to understand it.

> What if you create a grafana/ui component that has the style and the format you want and accept the text?

Enterprise datasources support 2 latest major Grafana versions. And even when G10 is released we might still support v8 for some time. If I add `@grafana/ui` component now, we will only be able to import it when we stop supporting v8 and switch to v9/v10 in enterprise datasources dependencies. This is an option of course, but not sure it's the best one. Also, there will still be duplication in each plugin's code which we wanted to avoid. We added `DatasourceDescription` component to `@grafana/plugin-ui` first, but then decided to move it to core.

Reply:
**Supporting earlier versions of Grafana**
I would say if you would like to do this with a plugin - which is totally fair - then it's the best to do it from the plugin, and not from the Grafana core / `"@grafan/ui"` side. The component (currently at least) looks relatively simple, so I would say it shouldn't be too bad to just keep it next to the plugin?

If the issue is sharing code between the different enterprise plugins as you say, did you consider for example having a monorepo for these plugins, where you can easily share these common bits?

Reply:
> The component (currently at least) looks relatively simple, so I would say it shouldn't be too bad to just keep it next to the plugin?

This is an option, but still not ideal I think, especially in case if we decide to make some change to this component. We have around 20 enterprise plugins... I understand this PR doesn't seem to get green light :) We will probably proceed with other option (probably reusing the component from [here](https://github.com/grafana/plugin-ui/blob/main/src/components/ConfigEditor/DataSourceDescription.tsx) in each plugin).

> If the issue is sharing code between the different enterprise plugins as you say, did you consider for example having a monorepo for these plugins, where you can easily share these common bits?

Yes, we have design doc for monorepo already.

Reply:
another long term consideration would be to extend DataSourcePlugin with something like 

```ts
export const plugin = new DataSourcePlugin(LokiDatasource)
  .setQueryEditor(LokiQueryEditorByApp)
  .setConfigEditorBuilder(builder => {
     builder.setHasRequiredFields(); 
     builder.addHttpSettings({..options})
     builder.addSection("Data links", sectionBuilder => {
        sectionBuilder.addStringInput({...} 
     })
   })
  .setQueryEditorHelp(LokiCheatSheet);
```

Something similar to panel options but with more customization and dynamic behaviors (looking at DS config they look a bit more dynamic, some requiring network requests to show/fill with options).

Reply:
What do you think if I move this `DataSourceDescription` component to `grafana/experimental` for now so that it can be used in both core and external plugins?